mmdet/utils/memory.py

```diff
@@ -0,0 +1,215 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+# Modified from https://github.com/facebookresearch/detectron2/blob/main/detectron2/utils/memory.py  # noqa
+import warnings
+from collections import abc
+from contextlib import contextmanager
+from functools import wraps
+
+import numpy as np
+import torch
+from mmcv.runner.fp16_utils import cast_tensor_type
+
+from mmdet.utils import get_root_logger
+
+
+def cast_tensor_device(inputs, src_device, dst_device):
+    """Recursively convert Tensor in inputs from src_device to dst_device.
+
+    Args:
+        inputs: Inputs that to be casted.
+        src_device (torch.device): Source device..
+        dst_device (torch.device): Destination device.
+
+    Returns:
+        The same device with inputs, but all contained Tensors have been cast.
+    """
+    if isinstance(inputs, torch.Tensor):
+        # we need to ensure that the type of inputs to be casted are the same
+        # as the argument `src_type`.
+        return inputs.to(dst_device) \
+            if inputs.device == src_device else inputs
```

Comment:
Why do we need to determine the source device?

Reply:
This logic is simply copied from mmcv.

Reply:
I consider it's unnecessary here to determine the source device. And, if we can make `cast_tensor_device` more general and can cast both tensor's device and type. So that, we can replace `maybe_to_cpu` and `maybe_to_fp16` with it.